README.md
# GraphQL For Serverless &middot;  [![NPM](https://img.shields.io/npm/v/graphql-serverless.svg?style=flat)](https://www.npmjs.com/package/graphql-serverless) [![Tests](https://travis-ci.org/nicolasdao/graphql-serverless.svg?branch=master)](https://travis-ci.org/nicolasdao/graphql-serverless) [![License](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg)](https://opensource.org/licenses/BSD-3-Clause) [![Neap](https://neap.co/img/made_by_neap.svg)](#this-is-what-we-re-up-to)  [![npm downloads](https://img.shields.io/npm/dt/graphql-serverless.svg?style=flat)](https://www.npmjs.com/package/graphql-serverless)
<a href="https://tracking.gitads.io/?repo=graphql-serverless" target="_blank"><img src="https://images.gitads.io/graphql-serverless" alt="GitAds"/></a>

__*graphql-serverless*__ is a middleware for [_**webfunc**_](https://github.com/nicolasdao/webfunc), that allows to deploy [GraphQL](http://graphql.org/learn/) apis (including an optional [GraphiQL interface](https://github.com/graphql/graphiql)) to the most popular serverless platforms. _**GraphQl Subscriptions**_ over websocket are also supported out-of-the-box (also supported in GraphiQL). Without changing a single line of code, seamlessly deploy to:
- [Zeit Now](https://zeit.co/now) (using express under the hood)
- [Google Cloud Functions](https://cloud.google.com/functions/) (incl. Firebase Function)
- [AWS Lambdas](https://aws.amazon.com/lambda)
- [Azure Functions](https://azure.microsoft.com/en-us/services/functions/) (COMING SOON...)

Copy/paste the following in your terminal if you want to run your first GraphQL api ([http://localhost:4000](http://localhost:4000)) including a GraphiQL interface ([http://localhost:4000/graphiql](http://localhost:4000/graphiql)) on your local machine in less than 30 seconds:

```
git clone https://github.com/nicolasdao/graphql-universal-server.git
cd graphql-universal-server
npm install
npm start 
```

This will serve 2 endpoints:

- [http://localhost:4000](http://localhost:4000): This is the GraphQL endpoint that your client can start querying.
- [http://localhost:4000/graphiql](http://localhost:4000/graphiql): This is the GraphiQL Web UI that you can use to test and query your GraphQL server. 

Deploying that api to [Zeit Now](https://zeit.co/now) will take between 15 seconds to 1.5 minute (depending on whether you need to login/creating a free Zeit account or not).

_If you haven't installed Zeit now-CLI yet or you need to login/create an account, then copy/paste this in your terminal:_
```
npm install now -g
now login
npm run deploy:prod
```

The above will work the exact same way whether you have an account or not. This is free, so don't worry about it.

_If you're already logged in, then simply run this:_
```
npm run deploy:prod
```

# Table Of Contents
> * [Install](#install)
> * [How To Use It](#how-to-use-it)
>   - [Basics](#basics)
>   - [GraphQl Subscriptions](#graphql-subscriptions)
>   - [Customizing GraphiQL](#customizing-graphiql)
>   - [Managing GraphQl Errors](#managing-graphql-errors)
>   - [Controling GraphQl Server Behavior With Custom Middleware](#controling-graphql-server-behavior-with-custom-middleware)
> * [About Us](#this-is-what-we-re-up-to)

# Install
```
npm install webfunc graphql-serverless --save
```

# How To Use It
## Basics
Using the template above (i.e. [graphql-universal-server](https://github.com/nicolasdao/graphql-universal-server.git)) is the easiest way to start a new GraphQL project from scratch. 

However, if you really want to start on a blank page:

1. Create a new npm project: `npm init`
2. Install the following: `npm install graphql-serverless webfunc --save`
3. Create an index.js as follow:

  ```js
  const { graphqlHandler, graphqlError } = require('graphql-serverless')
  const { makeExecutableSchema } = require('graphql-tools') // this dependency is automatically included in 'graphql-serverless'
  const { app } = require('webfunc')

  // STEP 1. Mock some data for this demo.
  const productMocks = [
    { id: 1, name: 'Product A', shortDescription: 'First product.' }, 
    { id: 2, name: 'Product B', shortDescription: 'Second product.' }]

  // STEP 2. Creating a basic GraphQl Schema.
  const schema = `
    type Product {
      id: ID!
      name: String!
      shortDescription: String
    }

    type Query {
      products(id: Int): [Product]
    }

    schema {
      query: Query
    }`

  const productResolver = {
    Query: {
      products(root, { id }, context) {
        const results = id ? productMocks.filter(p => p.id == id) : productMocks
        if (results.length > 0)
          return results
        else
          throw graphqlError(404, `Product with id ${id} does not exist.`)
      }
    }
  }

  const executableSchema = makeExecutableSchema({
    typeDefs: schema,
    resolvers: productResolver
  })

  // STEP 3. Creating a GraphQL and a GraphiQl endpoint
  const graphqlOptions = {
    schema: executableSchema,
    graphiql: { // If you do not want to host any GraphiQl web interface, leave this property undefined.
      endpoint: '/graphiql' 
    },
    context: {
    	someVar: 'This variable is passed in the "context" object in each resolver.'
    }
  }

  // Host a GraphQl API on 2 endpoints: '/' and '/graphiql'. '/graphiql' is used to host the GraphiQL web interface.
  // If you're not interested in the GraphiQl web interface, leave the above 'graphqlOptions.graphiql' undefined and 
  // replace the path following ['/', '/graphiql'] with '/'.
  app.all(['/', '/graphiql'], graphqlHandler(graphqlOptions))

  // STEP 4. Starting the server 
  eval(app.listen('app', 4000))
  ```

4. Then simply run:
  ```
  node index.js
  ```

This will serve 2 endpoints:

- [http://localhost:4000](http://localhost:4000): This is the GraphQL endpoint that your client can start querying.
- [http://localhost:4000/graphiql](http://localhost:4000/graphiql): This is the GraphiQL Web UI that you can use to test and query your GraphQL server. 

>If you need best practices on how to structure your GraphQL project, clone the [graphql-universal-server](https://github.com/nicolasdao/graphql-universal-server.git) project and see by yourself. 

## GraphQl Subscriptions

> __WARNING: This feature is only available on [Zeit Now serverless](https://zeit.co/now) or on localhost.__ 
> Even though _graphql-serverless_ relies on [_webfunc_](https://github.com/nicolasdao/webfunc) to deploy on FaaS solutions like AWS Lambdas or Google Functions, because those hosting platforms do not natively support websocket, GraphQl Subscriptions can't be deployed there.

<p align="center"><img src="https://raw.githubusercontent.com/nicolasdao/graphql-serverless/master/img/graphql_subscription.gif"/></p>

_graphql-serverless_ exposes a helper method `setupSubscriptions` that can host a websocket endpoint for GraphQl Subscriptions. In the following example, we will slightly modify the code above to:
- (MODIFICATION A) Configure a new websocket endpoint for all subscriptions.
- (MODIFICATION B) Add a PubSub queue so that publisher can publish messages onto topics and subscribers can listen to certain topics so that clients using websocket can receive updates.
- (MODIFICATION C) Add a new GraphQl Mutation to insert a new product. This insert will act as a publisher. It will add a message to the PubSub topic once the product has been successfully inserted.
- (MODIFICATION D) Add a new GraphQl Subscription that listen to a specific topic on the PubSub queue and uses websocket to inform the client that a new product has been inserted.

Install _graphql-subscriptions_: 

`npm install graphql-subscriptions --save`

Update the previous `index.js` as follow:
  ```js
  // MODIFICATION A - Import the 'setupSubscriptions' helper
  const { graphqlHandler, graphqlError, setupSubscriptions } = require('graphql-serverless')

  ...

  // MODIFICATION B - Create a simple local pub/sub (not scalable option, but good enough for a demo) 
  const { PubSub } = require('graphql-subscriptions')
  const pubsub = new PubSub()

  ...

  // MODIFICATION C/D - Add an 'insert product' MUTATION and a 'product inserted' SUBSCRIPTION in the GraphQl schema
  const schema = `
  input NewProductInput {
    name: String!
    shortDescription: String
  }

  type Mutation {
    productInsert(product: NewProductInput!): Product
  }

  type Subscription {
    productInserted: Product
  }

  schema {
    query: Query
    mutation: Mutation
    subscription: Subscription
  }`

  ...

  // MODIFICATION C/D - Add an 'insert product' MUTATION and a 'product inserted' SUBSCRIPTION in the GraphQl product resolver
  const productResolver = {
    Query: {...},

    Mutation: {
      productInsert(root, { product }, context) {
        if (!product || !product.name)
          throw context.graphqlError('Missing required argument \'product.name\'.')

        const newId = productMocks.sort((a,b) => a.id < b.id)[0].id + 1
        const newProduct = Object.assign({ id: newId }, product)
        productMocks.push(newProduct)
        pubsub.publish('productInserted', { productInserted: newProduct })
        return newProduct
      }
    },

    Subscription: {
      productInserted: {
        subscribe: () => pubsub.asyncIterator('productInserted')
      }
    }
  }

  ...

  // MODIFICATION A - Define the location of the subscriptions endpoint
  const graphqlOptions = {
    schema: executableSchema,
    graphiql: {
      endpoint: '/graphiql'
    },
    subscriptionsEndpoint: '/subscriptions' // this means that the subscription endpoint is 'ws://localhost:4000/subscriptions' if you're deploying locally
  }

  ...

  // MODIFICATION A - Start the websocket endpoint after the server as started. 
  // WARNING: This only works for localhost, serverless Zeit Now, but not 
  // for FaaS like AWS Lambdas, Google Functions, ...
  eval(app.listen('app', 4000, () => setupSubscriptions(app.server, graphqlOptions)))
  ```

Execute `node index.js` and then browse to [http://localhost:4000/graphiql](http://localhost:4000/graphiql). Start a subscription as follow:
```js
subscription {
  productInserted {
    id
    name
  }
}
```

At that point, the client is simply listening to any new messages on the 'productInserted' topic. Time to publish a new messages on that topic. Open a new tab and browse again to [http://localhost:4000/graphiql](http://localhost:4000/graphiql). There insert a new product as follow:
```js
mutation {
  productInsert(product: {
    name: "Product C"
  }) {
    id
    name
  }
}
```

Once the product has been inserted, you should be able to observe that your subscription client has noticed it. 

<p align="center"><img src="https://raw.githubusercontent.com/nicolasdao/graphql-serverless/master/img/graphql_subscription.gif"/></p>

## Customizing GraphiQL

The code sample in the [__*Basics*__](#basics) section uses the default GraphiQl settings:

<p align="center"><img src="https://raw.githubusercontent.com/nicolasdao/graphql-serverless/master/img/graphiql_light_theme.png" width="600"/></p>

By updating the `graphqlOptions` in the [__*Basics*__](#basics) section example as follow:

```js
const graphqlOptions = {
	schema: executableSchema,
	graphiql: { 
		endpoint: '/graphiql',
		head: {
			title: 'Neap GraphQl API',
			// Adding a custom Favicon
			custom: '<link rel="shortcut icon" href="https://neap.co/favicon.ico">',
			// Change the default 'light' theme to a dark one.
			theme: 'dark',
			// Replace the default 'GraphiQl' logo to your own
			logo: '<div class="title"><img src="https://neap.co/img/neap_white_small_logo.png" style="width: 88px;z-index: 7;padding-left: 24px;"></div>'
		},    
		// Adding a custom JS script
		script: () => {
			function getCookie(cname) {
				var name = cname + '='
				var decodedCookie = decodeURIComponent(document.cookie)
				var ca = decodedCookie.split(';')
				for(var i = 0; i <ca.length; i++) {
					var c = ca[i]
					while (c.charAt(0) == ' ')
						c = c.substring(1)
					if (c.indexOf(name) == 0)
						return c.substring(name.length, c.length)
				}
				return ''
			}
		},
		// Executing a custom JS function each time a GraphQL request is made
		onRequest: headers => {
			var token = getCookie('neap_cookie')
			if (token)
				headers.Authorization = 'bearer ' + token
		}
	}
}
```

We can update the GraphiQL interface as follow:

<p align="center"><img src="https://raw.githubusercontent.com/nicolasdao/graphql-serverless/master/img/graphiql_dark_theme.png" width="600"/></p>

The differences are:
- Custom page name
- Custom favicon
- Changing from the default 'light' theme to a dark one
- Custom logo
- Custom javascript 
- Custom function running each time a GraphQl request is made. In our case, we're updating the HTTP headers to add a bearer token stored in a 'neap_cookie' cookie.

To use a custom CSS rather than relying on the 'light' or 'dark' theme, you can use the `css` property:

```js
const graphqlOptions = {
  schema: executableSchema,
  graphiql: { 
    endpoint: '/graphiql',
    head: {
      // Adding a dark theme
      css: [
        'https://neap.co/resources/css/graphiql/0.0.1/dark_style.css', 
        'https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700|Source+Code+Pro:200,400,700']
    }
  }
}
```

> WARNING: Using the `css` property will override completely any other out-of-the-box css.

## Managing GraphQl Errors

By default, any uncaught errors are marshalled to the graphql response similar to this:
```js
{
  "errors": [
    {
      "message": "Product with id 20 does not exist.",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "products"
      ]
    }
  ],
  "data": {
    "products": null
  }
}
```

This type of uncaught error also yield a 500 HTTP code. A piece of code that could produce the error above could be:
```js
const productResolver = {
  Query: {
    products(root, { id }, context) {
      const results = id ? productMocks.filter(p => p.id == id) : productMocks

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.9.0](https://github.com/nicolasdao/google-graphql-functions/compare/v0.8.3...v0.9.0) (2020-07-23)


### Features

* Add better error handling support ([f568e15](https://github.com/nicolasdao/google-graphql-functions/commit/f568e15f953539029b7bf18bea781ca30bd3336d))

<a name="0.8.3"></a>
## [0.8.3](https://github.com/nicolasdao/google-graphql-functions/compare/v0.8.2...v0.8.3) (2019-09-24)


### Bug Fixes

* The server fails with no proper error message when the graphql query is malformed. ([009677c](https://github.com/nicolasdao/google-graphql-functions/commit/009677c))



<a name="0.8.2"></a>
## [0.8.2](https://github.com/nicolasdao/google-graphql-functions/compare/v0.8.1...v0.8.2) (2018-11-30)



<a name="0.8.1"></a>
## [0.8.1](https://github.com/nicolasdao/google-graphql-functions/compare/v0.8.0...v0.8.1) (2018-07-06)


### Bug Fixes

* The GraphiQL dark theme caret is almost invisible. Make it brighter. ([17b3297](https://github.com/nicolasdao/google-graphql-functions/commit/17b3297))



<a name="0.8.0"></a>
# [0.8.0](https://github.com/nicolasdao/google-graphql-functions/compare/v0.7.8...v0.8.0) (2018-05-06)



<a name="0.7.8"></a>
## [0.7.8](https://github.com/nicolasdao/google-graphql-functions/compare/v0.7.7...v0.7.8) (2018-05-06)



<a name="0.7.7"></a>
## [0.7.7](https://github.com/nicolasdao/google-graphql-functions/compare/v0.7.6...v0.7.7) (2018-05-06)



<a name="0.7.6"></a>
## [0.7.6](https://github.com/nicolasdao/google-graphql-functions/compare/v0.7.5...v0.7.6) (2018-05-06)


### Bug Fixes

* Incompatible with latest version of graphql-s2s because mismatch graphql version. Please downgrade to graphql 0.11.0 ([e2aafa5](https://github.com/nicolasdao/google-graphql-functions/commit/e2aafa5))



<a name="0.7.5"></a>
## [0.7.5](https://github.com/nicolasdao/google-graphql-functions/compare/v0.7.4...v0.7.5) (2018-03-17)



<a name="0.7.4"></a>
## [0.7.4](https://github.com/nicolasdao/google-graphql-functions/compare/v0.7.3...v0.7.4) (2018-03-03)



<a name="0.7.3"></a>
## [0.7.3](https://github.com/nicolasdao/google-graphql-functions/compare/v0.7.2...v0.7.3) (2018-03-03)


### Bug Fixes

* GraphiQl theme is broken if it is not explicitely defined in the graphqlOptions ([f37cf5e](https://github.com/nicolasdao/google-graphql-functions/commit/f37cf5e))



<a name="0.7.2"></a>
## [0.7.2](https://github.com/nicolasdao/google-graphql-functions/compare/v0.7.1...v0.7.2) (2018-03-03)


### Bug Fixes

* CSS issue in graphiQL ([5e57d9d](https://github.com/nicolasdao/google-graphql-functions/commit/5e57d9d))



<a name="0.7.1"></a>
## [0.7.1](https://github.com/nicolasdao/google-graphql-functions/compare/v0.7.0...v0.7.1) (2018-03-02)



<a name="0.7.0"></a>
# [0.7.0](https://github.com/nicolasdao/google-graphql-functions/compare/v0.6.1...v0.7.0) (2018-03-02)


### Features

* Add support for custom logo ([47a93dc](https://github.com/nicolasdao/google-graphql-functions/commit/47a93dc))



<a name="0.6.1"></a>
## [0.6.1](https://github.com/nicolasdao/google-graphql-functions/compare/v0.6.0...v0.6.1) (2018-02-28)



<a name="0.6.0"></a>
# [0.6.0](https://github.com/nicolasdao/google-graphql-functions/compare/v0.5.0...v0.6.0) (2018-02-28)



<a name="0.5.0"></a>
# [0.5.0](https://github.com/nicolasdao/google-graphql-functions/compare/v5.0.1...v0.5.0) (2018-02-25)



<a name="5.0.1"></a>
## [5.0.1](https://github.com/nicolasdao/google-graphql-functions/compare/v5.0.0...v5.0.1) (2018-02-25)


### Bug Fixes

* Secure websocket not supported in GraphiQl ([4002050](https://github.com/nicolasdao/google-graphql-functions/commit/4002050))



<a name="5.0.0"></a>
# [5.0.0](https://github.com/nicolasdao/google-graphql-functions/compare/v0.4.1...v5.0.0) (2018-02-24)


### Features

* Add new 'graphqlError' function that allows to create better error messages including http code ([dc60883](https://github.com/nicolasdao/google-graphql-functions/commit/dc60883))
* Add support for subscription ([07f3167](https://github.com/nicolasdao/google-graphql-functions/commit/07f3167))



<a name="0.4.1"></a>
## [0.4.1](https://github.com/nicolasdao/google-graphql-functions/compare/v0.4.0...v0.4.1) (2018-02-20)



<a name="0.4.0"></a>
# [0.4.0](https://github.com/nicolasdao/google-graphql-functions/compare/v0.4.0-alpha.12...v0.4.0) (2018-02-20)



<a name="0.4.0-alpha.12"></a>
# [0.4.0-alpha.12](https://github.com/nicolasdao/google-graphql-functions/compare/v0.4.0-alpha.11...v0.4.0-alpha.12) (2018-02-19)


### Features

* Upgrade to latest graphql dependencies ([b160bea](https://github.com/nicolasdao/google-graphql-functions/commit/b160bea))



<a name="0.4.0-alpha.11"></a>
# [0.4.0-alpha.11](https://github.com/nicolasdao/google-graphql-functions/compare/v0.4.0-alpha.10...v0.4.0-alpha.11) (2018-02-13)


### Bug Fixes

* onResponse and request.graphql.transform are not failing properly which causes unclear errors as well as affecting the overall server performance by blocking for 2 seconds ([c389454](https://github.com/nicolasdao/google-graphql-functions/commit/c389454))



<a name="0.4.0-alpha.10"></a>
# [0.4.0-alpha.10](https://github.com/nicolasdao/google-graphql-functions/compare/v0.4.0-alpha.9...v0.4.0-alpha.10) (2018-02-11)


### Features

* Add support for optionally transform the graphql result ([b3e3898](https://github.com/nicolasdao/google-graphql-functions/commit/b3e3898))



<a name="0.4.0-alpha.9"></a>
# [0.4.0-alpha.9](https://github.com/nicolasdao/google-graphql-functions/compare/v0.4.0-alpha.8...v0.4.0-alpha.9) (2018-01-30)



<a name="0.4.0-alpha.8"></a>
# [0.4.0-alpha.8](https://github.com/nicolasdao/google-graphql-functions/compare/v0.4.0-alpha.7...v0.4.0-alpha.8) (2018-01-30)


### Features


src/index.js
/**
 * Copyright (c) 2018, Neap pty ltd.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * This file incorporates work covered by the following copyright and  
 * permission notice:
 *
 *    Copyright (c) 2015, Facebook, Inc.
 *    All rights reserved.
 *
 *    This source code is licensed under the BSD-style license found in the
 *    LICENSE file in the root directory of this source tree. An additional grant
 *    of patent rights can be found in the PATENTS file in the same directory.
 */

const accepts = require('accepts')
const graphql = require('graphql')
const httpError = require('http-errors')
const url = require('url')
const { SubscriptionServer } = require('subscriptions-transport-ws')
const graphqlError = require('./utils')
require('color')

const parseBody = require('./parseBody')
const { renderGraphiQL } = require('./renderGraphiQL')

const graphqlHandler = options => {
	if (!options)
		throw new Error('Missing required argument. A graphql handler must accept an \'options\' object')

	return (req, res, next) => {
		if (!res.headersSent) {
			const httpHandler = graphqlHTTP(options)
			if (!httpHandler)
				throw httpError(500, 'GraphQL middleware requires a valid \'getOptions\' argument.')
			return httpHandler(req, res).then(() => next())
		}
		else
			next()
	}
}

// Though this may seem completely stupid to list a combination of potential empty queries rather than 
// escaping the original query, empty queries are such an exception that escaping all freaking 
// queries just for that rare occasion seems too expensive.
const _basicEmptyQueries = {'{}': true,'{\n}': true,'{ \n}': true,'{ }': true,'query{}': true,'query{\n}': true,'query{ \n}': true,'query{ }': true,'query {}': true,'query {\n}': true,'query { \n}': true,'query { }': true,'mutation{}': true,'mutation{\n}': true,'mutation{ \n}': true,'mutation{ }': true,'mutation {}': true,'mutation {\n}': true,'mutation { \n}': true,'mutation { }': true,'subscription{}': true,'subscription{\n}': true,'subscription{ \n}': true,'subscription{ }': true,'subscription {}': true,'subscription {\n}': true,'subscription { \n}': true,'subscription { }': true}
function graphqlHTTP(options) {
	if (!options) {
		throw new Error('GraphQL middleware requires options.')
	}
	
	return (request, response) => {
		// Higher scoped variables are referred to at various stages in the
		// asynchronous state machine below.
		let schema
		let context
		let rootValue
		let pretty
		let graphiql
		let formatErrorFn
		let extensionsFn
		let showGraphiQL
		let query
		let documentAST
		let variables
		let operationName
		let validationRules
		let endpointURL
		let subscriptionsEndpoint
		let websocketConnectionParams
		let schemaAST
		let graphiQlOptions = {}
		let onResponse = (req, res, result) => Promise.resolve(result)

		// Promises are used as a mechanism for capturing any thrown errors during
		// the asynchronous process below.

		// Resolve the Options to get OptionsData.
		return new Promise(resolve => {
			resolve(
				typeof options === 'function' ?
					options(request, response) :
					options
			)
		}).then(optionsData => {
			// Assert that optionsData is in fact an Object.
			if (!optionsData || typeof optionsData !== 'object') {
				throw new Error(
					'GraphQL middleware option function must return an options object ' +
																				'or a promise which will be resolved to an options object.'
				)
			}

			// Assert that schema is required.
			if (!optionsData.schema) {
				throw new Error(
					'GraphQL middleware options must contain a schema.'
				)
			}

			// Collect information from the options data object.
			schema = optionsData.schema
			context = optionsData.context || request
			rootValue = optionsData.rootValue
			pretty = optionsData.pretty
			graphiql = optionsData.graphiql ? true : false
			formatErrorFn = optionsData.formatError
			extensionsFn = optionsData.extensions
			endpointURL = optionsData.endpointURL
			subscriptionsEndpoint = optionsData.subscriptionsEndpoint
			websocketConnectionParams = optionsData.websocketConnectionParams
			schemaAST = optionsData.schemaAST
			if (optionsData.graphiql != undefined && typeof(optionsData.graphiql) == 'object')
				graphiQlOptions =  optionsData.graphiql
			if (graphiQlOptions.toggle == false)
				graphiql = false
			if (graphiQlOptions.endpoint)
				endpointURL = graphiQlOptions.endpoint
			if (optionsData.onResponse && typeof(optionsData.onResponse) == 'function')
				onResponse = (req, res, result) => Promise.resolve(null).then(() => optionsData.onResponse(req, res, result))

			validationRules = graphql.specifiedRules
			if (optionsData.validationRules) {
				validationRules = validationRules.concat(optionsData.validationRules)
			}

			// GraphQL HTTP only supports GET and POST methods.
			if (request.method !== 'GET' && request.method !== 'POST') {
				response.setHeader('Allow', 'GET, POST')
				throw httpError(405, 'GraphQL only supports GET and POST requests.')
			}

			// Parse the Request to get GraphQL request parameters.
			return getGraphQLParams(request)
		}).then(params => {
			// Get GraphQL params from the request and POST body data.
			query = params.query
			if (!_basicEmptyQueries[query]) {
				variables = params.variables
				operationName = params.operationName && params.operationName != 'undefined' ? params.operationName : null
				showGraphiQL = graphiql && canDisplayGraphiQL(request, params)

				if (showGraphiQL && endpointURL) {
					const pathname = (p => p ? p : '')(url.parse(request.url).pathname).replace(/^\//, '').toLowerCase()
					const endpointsParts = endpointURL.toLowerCase().split('/').filter(x => x && x != '/')
					const pathnameLastParts = pathname.split('/').filter(x => x && x != '/').slice(-endpointsParts.length)
					const isGraphiQlRequest = pathnameLastParts.join('_._') == endpointsParts.join('_._')
					
					if (!isGraphiQlRequest)
						showGraphiQL = false
				}

				// If there is no query, but GraphiQL will be displayed, do not produce
				// a result, otherwise return a 400: Bad Request.
				if (!query) {
					if (showGraphiQL) {
						return null
					}
					throw httpError(400, 'Must provide query string.')
				}

				// GraphQL source.
				const source = new graphql.Source(query, 'GraphQL request')

				// Parse source to AST, reporting any syntax error.
				try {
					documentAST = graphql.parse(source)
				} catch (syntaxError) {
					// Return 400: Bad Request if any syntax errors errors exist.
					response.statusCode = 400
					return { errors: [syntaxError] }
				}

				// Validate AST, reporting any errors.
				const validationErrors = graphql.validate(schema, documentAST, validationRules)
				if (validationErrors.length > 0) {
					// Return 400: Bad Request if any validation errors exist.
					response.statusCode = 400
					return { errors: validationErrors }
				}

				// Only query operations are allowed on GET requests.
				if (request.method === 'GET') {
					// Determine if this GET request will perform a non-query.
					const operationAST = graphql.getOperationAST(documentAST, operationName)
					if (operationAST && operationAST.operation !== 'query') {
						// If GraphiQL can be shown, do not perform this query, but
						// provide it to GraphiQL so that the requester may perform it
						// themselves if desired.
						if (showGraphiQL) {
							return null
						}

						// Otherwise, report a 405: Method Not Allowed error.
						response.setHeader('Allow', 'POST')
						throw httpError(
							405,
							`Can only perform a ${operationAST.operation} operation ` +
																									'from a POST request.'
						)
					}
				}
				// Perform the execution, reporting any errors creating the context.
				try {
					return graphql.execute(
						schema,
						documentAST,
						rootValue,
						context,
						variables,
						operationName
					)
				} catch (contextError) {
					// Return 400: Bad Request if any execution context errors exist.
					response.statusCode = 400
					return { errors: [contextError] }
				}
			}
			else
				return { ___empty: true, data: {} }
		}).then(result => {
			// Collect and apply any metadata extensions if a function was provided.
			// http://facebook.github.io/graphql/#sec-Response-Format
			if (result && extensionsFn && !result.___empty) {
				return Promise.resolve(extensionsFn({
					document: documentAST,
					variables,
					operationName,
					result
				})).then(extensions => {
					if (extensions && typeof extensions === 'object') {
						(result).extensions = extensions
					}
					return result
				})
			}
			return result
		}).catch(error => {
			// If an error was caught, report the httpError status, or 500.
			response.statusCode = error.status || 500
			return { errors: [error] }
		}).then(result => {
			let httpCode = response.statusCode || 200
			if (result && result.___empty) {
				result.errors = null
				httpCode == 200
				result = {}
			}
			let transformResult = () => Promise.resolve(result)

			if (result) {
				if (request.graphql) {
					// Add custom errors from potential middleware.
					if (request.graphql.errors && Array.isArray(request.graphql.errors) && request.graphql.errors.length > 0) {
						if (!result.errors)
							result.errors = []
						result.errors.push(...request.graphql.errors.map(({ message, locations, path }) => ({ message, locations, path })))
					}
					
					// Add custom warnings from potential middleware.
					if (request.graphql.warnings && Array.isArray(request.graphql.warnings) && request.graphql.warnings.length > 0) {
						if (!result.warnings)
							result.warnings = []
						result.warnings.push(...request.graphql.warnings.map(({ message, locations, path }) => ({ message, locations, path })))
					}

					// Transform final response based on potential middleware rules.
					if (request.graphql.transform)
						transformResult = r => Promise.resolve(null).then(() => request.graphql.transform(r))
				}

				// Format any encountered errors.
				if (result.errors && result.errors.length > 0) {
					const formattingError = formatErrorFn || graphql.formatError
					let explicitHttpCode
					result.errors = result.errors.reduce((acc, e) => {
						// 'newError' example: 
						// {
						//		message: '{"type":"graphql","httpCode":422,"message":"Failed to do something","alternateMessage":"Internal Server Error"}',
						//		locations: [ { line: 2, column: 3 } ],
						//		path: [ 'conditionUpdate' ]
						// }
						const newError = formattingError(e)
						// The block of code below reformats the 'newError.message' property
						try {
							const customError = JSON.parse(newError.message) || {}
							if (customError.type == 'graphql') {
								explicitHttpCode = customError.httpCode || 500
								if (customError.errors && Array.isArray(customError.errors))
									acc.push(...customError.errors)
								else
									acc.push({
										message:customError.hideErrors && customError.alternateMessage ? customError.alternateMessage : customError.message
									})
								
								return acc
							}
						} catch (err) {
							(() => null)(err)
						}

						acc.push(newError)
						return acc
					}, [])
					if (!explicitHttpCode && allPropertiesFalsy(result.data))
						httpCode = response.statusCode < 500 ? 500 : response.statusCode
					else if (explicitHttpCode)
						httpCode = explicitHttpCode
				}
			}
			
			const execute = showGraphiQL
				// If allowed to show GraphiQL, present it instead of JSON.
				? () => {
					const payload = renderGraphiQL({ query, variables, operationName, result, schemaAST, subscriptionsEndpoint, websocketConnectionParams }, graphiQlOptions)
					response.setHeader('Content-Type', 'text/html; charset=utf-8')
					sendResponse(response, payload, httpCode)
					return result
				}
				// Otherwise, present JSON directly. 
				: () => onResponse(request, response, result).catch(err => ({ _error: err, _location: 'Function \'optionsData.onResponse\'', _result: result }))
					.then(r => r._error ? r : transformResult(r).catch(err => ({ _error: err, _location: 'Function \'request.graphql.transform\'', _result: r }))).then(r => {
						let _result = r || result
						if (r && r._error) {
							_result = r._result
							if (!_result.errors) 
								_result.errors = []
							_result.errors.push({ message: `${r._error.message}\n${r._error.stack}`, location: r._location })
							httpCode = 500
						}

						const payload = JSON.stringify(_result, null, pretty ? 2 : 0)
						response.setHeader('Content-Type', 'application/json; charset=utf-8')
						sendResponse(response, payload, httpCode)
						return _result
					})
			
			return execute()
		}).catch(err => {
			console.log(err.stack)
			throw err
		})
	}
}

const allPropertiesFalsy = data => {
	let result = true
	if (data && typeof(data) == 'object')
		for (let k in data) {
			result = result && !data[k]
		}
	return result
}

const getParseDataFromUrl = req => {
	if (req.url){
		try {
			return url.parse(req.url, true).query
		}
		/*eslint-disable */
		catch(err) {
			return {}
		}
		/*eslint-enable */
	}
	else
		return {}
}

/**
 * Provided a "Request" provided by express or connect (typically a node style
 * HTTPClientRequest), Promise the GraphQL request parameters.
 */
function getGraphQLParams(request) {
	return parseBody(request).then(bodyData => {
		const data =
			request.graphql && request.graphql.query ? request.graphql :
				bodyData && bodyData.query ? bodyData : getParseDataFromUrl(request)

		return parseGraphQLParams(data)
	})
}

/**
 * Helper function to get the GraphQL params from the request.
 */
function parseGraphQLParams(data={}) {
	// GraphQL Query string.
	let query = data.query
	if (typeof query !== 'string') {
		query = null
	}

	// Parse the variables if needed.
	let variables = data.variables
	if (variables && typeof variables === 'string') {
		try {
			variables = JSON.parse(variables)
		} catch (error) {
			throw httpError(400, 'Variables are invalid JSON.')
		}
	} else if (typeof variables !== 'object') {
		variables = null
	}
	// Name of GraphQL operation to execute.
	let operationName = data.operationName
	if (typeof operationName !== 'string') {
		operationName = null
	}
	const raw = data.raw !== undefined

	return { query, variables, operationName, raw }
}

/**
 * Helper function to determine if GraphiQL can be displayed.
 */
function canDisplayGraphiQL(request, params) {
	// If `raw` exists, GraphiQL mode is not enabled.
	// Allowed to show GraphiQL if not requested as raw and this request
	// prefers HTML over JSON.
	return !params.raw && accepts(request).types(['json', 'html']) === 'html'
}

/**
 * Helper function for sending the response data. Use response.send it method
 * exists (express), otherwise use response.end (connect).
 */
function sendResponse(response, data, httpCode=200) {
	if (typeof response.send === 'function') {
		response.status(httpCode).send(data)
	} else {
		response.statusCode = httpCode
		response.end(data)
	}
}

const setupSubscriptions = (server, { schema, subscriptionsEndpoint }) => {
	if (!schema)
		throw new Error('\'schema\' is required.')
	if (!subscriptionsEndpoint)
		throw new Error('\'subscriptionsEndpoint\' is required.')
	const subServer = new SubscriptionServer({
		execute: graphql.execute,
		subscribe: graphql.subscribe,
		schema: schema
	}, {
		server,
		path: subscriptionsEndpoint.indexOf('ws:') == 0 ? subscriptionsEndpoint : `/${subscriptionsEndpoint.replace(/^\/*/, '')}`
	})
	console.log('WebSocket successfully set up for handling GraphQL subscriptions.'.cyan)
	return subServer
}

const isGraphiQLRequest = (req) => getGraphQLParams(req).then(params => canDisplayGraphiQL(req, params))

module.exports = {
	graphqlHandler,
	isGraphiQLRequest,
	graphqlError,
	setupSubscriptions
}